src/entity/Post.ts

```diff
@@ -179,3 +184,238 @@ export const getAuthorPostStats = (
     .from(Post, 'post')
     .where({ authorId })
     .getRawOne<PostStats>();
+
+export interface AddPostData {
+  id: string;
+  title: string;
+  url: string;
+  publicationId: string;
+  publishedAt?: string | Date;
+  createdAt?: Date;
+  image?: string;
+  ratio?: number;
+  placeholder?: string;
+  tags?: string[];
+  siteTwitter?: string;
+  creatorTwitter?: string;
+  readTime?: number;
+  canonicalUrl?: string;
+  keywords?: string[];
+  description?: string;
+  toc?: Toc;
+  summary?: string;
+}
+
+const parseReadTime = (
+  readTime: number | string | undefined,
+): number | undefined => {
+  if (!readTime) {
+    return undefined;
+  }
+  if (typeof readTime == 'number') {
+    return Math.floor(readTime);
+  }
+  return Math.floor(parseInt(readTime));
+};
+
+type Reason = 'exists' | 'author banned' | 'missing fields';
+export type AddNewPostResult =
+  | { status: 'ok'; postId: string }
+  | { status: 'failed'; reason: Reason; error?: Error };
+
+const checkRequiredFields = (data: AddPostData): boolean => {
+  return !!(data && data.title && data.url && data.publicationId);
+};
+
+const shouldAddNewPost = async (
+  con: Connection,
+  data: AddPostData,
+): Promise<Reason | null> => {
+  const p = await con
+    .getRepository(Post)
+    .createQueryBuilder()
+    .select('id')
+    .where(
+      'url = :url or url = :canonicalUrl or "canonicalUrl" = :url or "canonicalUrl" = :canonicalUrl',
+      { url: data.url, canonicalUrl: data.canonicalUrl },
+    )
+    .getRawOne();
+  if (p) {
+    return 'exists';
+  }
+  if (data.creatorTwitter === '@NewGenDeveloper') {
+    return 'author banned';
+  }
+
+  if (!data.title) {
+    return 'missing fields';
+  }
+};
+
+const fixAddPostData = (data: AddPostData): AddPostData => ({
+  ...data,
+  id: shortid.generate(),
+  canonicalUrl: data.canonicalUrl || data.url,
+  title: data.title && he.decode(data.title),
+  createdAt: new Date(),
+  readTime: parseReadTime(data.readTime),
+  creatorTwitter:
+    data.creatorTwitter === '' || data.creatorTwitter === '@'
+      ? null
+      : data.creatorTwitter,
+  publishedAt: data.publishedAt && new Date(data.publishedAt),
+});
+
+const mergeKeywords = async (
+  entityManager: EntityManager,
+  keywords?: string[],
+): Promise<{ mergedKeywords: string[]; allowedKeywords: string[] }> => {
+  if (keywords?.length) {
+    const synonymKeywords = await entityManager.getRepository(Keyword).find({
+      where: {
+        status: 'synonym',
+        value: In(keywords),
+      },
+    });
+    const mergedKeywords = Array.from(
```

Comment:
In the apps we refactor something similar to more named version.
So the Array.from new Set could be something like uniqueRecords.

Not urgent, but would maybe be good to generalize these things along our code bases.
https://github.com/dailydotdev/apps/blob/209ad77c5801c8475bfb0c38cfafc4c683acade1/packages/eslint-rules/lib/rules/no-custom-color.js#L44

Reply:
Great idea!